File: server/src/services/composition_service.ts
import { PrismaClient } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';

const prisma = new PrismaClient();

interface CompositionItem {
    id: number;
    title: string;
    type: string;
    content: string;
    examples?: string;
    theme?: string;
    isFree: number;
}

async function list(isFree?: boolean): Promise<CompositionItem[]> {
    const where: any = { status: 1 };
    if (isFree !== undefined) where.isFree = isFree ? 1 : 0;

    const items = await prisma.compositionTemplate.findMany({
        where,
        orderBy: { id: 'asc' },
    });

    return items.map((i) => ({
        id: i.id,
        title: i.title,
        type: i.type,
        content: i.content,
        examples: i.examples ?? undefined,
        theme: i.theme ?? undefined,
        isFree: i.isFree,
    }));
}

async function getById(id: number): Promise<CompositionItem | null> {
    const item = await prisma.compositionTemplate.findUnique({ where: { id } });
    if (!item || item.status !== 1) return null;

    return {
        id: item.id,
        title: item.title,
        type: item.type,
        content: item.content,
        examples: item.examples ?? undefined,
        theme: item.theme ?? undefined,
        isFree: item.isFree,
    };
}

export { list, getById };

File: server/app/services/composition_service.py
"""作文模板服务（对应 TS composition_service.ts）"""
from sqlalchemy.orm import Session

from ..models import CompositionTemplate


def _toItem(i: CompositionTemplate) -> dict:
    return {
        "id": i.id,
        "title": i.title,
        "type": i.type,
        "content": i.content,
        "examples": i.examples,
        "theme": i.theme,
        "isFree": i.isFree,
    }


def list(db: Session, isFree: bool | None = None) -> list:
    query = db.query(CompositionTemplate).filter(CompositionTemplate.status == 1)
    if isFree is not None:
        query = query.filter(CompositionTemplate.isFree == (1 if isFree else 0))
    items = query.order_by(CompositionTemplate.id.asc()).all()
    return [_toItem(i) for i in items]


def getById(db: Session, id: int) -> dict | None:
    item = db.query(CompositionTemplate).filter(CompositionTemplate.id == id).first()
    if not item or item.status != 1:
        return None
    return _toItem(item)
